src/btree/bt_delete.c

```diff
@@ -314,8 +314,8 @@ __tombstone_update_alloc(WT_SESSION_IMPL *session,
 {
 	WT_UPDATE *upd;
 
-	WT_RET(
-	    __wt_update_alloc(session, NULL, &upd, sizep, WT_UPDATE_TOMBSTONE));
+	WT_RET(__wt_update_alloc(session,
+	    NULL, &upd, sizep, WT_UPDATE_TOMBSTONE, WT_PREPARE_INIT));
```

Comment:
Instead of setting prepare_state to WT_PREPARE_INIT, should I be checking for ```ref->page_del->prepare_state``` in the caller ```__wt_delete_page_instantiate()``` and use that instead ? This concern is coming from line 369, considering that it is possible to have a deleted prepared page ?
I suspect not because I do not see ```upd->prepare_state``` getting set around this code.

Reply:
The value of WT_PREPARE_INIT is choosed as '0' to avoid it setting specifically.  __wt_update_alloc will initialize to zero. My preference is not to modify __wt_update_alloc signature. A page with prepared update cannot be freed and hence page will not be instantiated.

Reply:
When reading an update from lookaside we need to pass the corresponding prepare_state. Hence we need to change the signature of __wt_update_alloc(). eg:
https://github.com/wiredtiger/wiredtiger/pull/4215/files#diff-e80f0967af0f71afc192ac61c363333cR187

What I have a minor doubt is when we are re-initiating a deleted page with __wt_delete_page_instantiate(), from the code at bt_delete.c:369 it looks like the deleted page can be in prepared state. Question I have over there, that in such a case, would the updates for the records on that page be in the same prepare_state as the page itself or be default initialised to WT_PREPARE_INIT. I think they will be default initialised to WT_PREPARE_INIT, because that is what happens today and I don't see how long running prepared will change that.

Reply:
I mean instead of modify __wt_update_alloc signature, where ever required we can update the prepared state of update structure after it is allocated.

Reply:
If a page is deleted, it will have page_del structure, and prepare state will be set to page_del structure. It will not have updates.

Reply:
> I mean instead of modify __wt_update_alloc signature, where ever required we can update the prepared state of update structure after it is allocated

I will like to keep it as I have in the change at the moment, ie to pass the prepare_state through the __wt_update_alloc() call. I think this is more consistent with the notion that we pass all the needed information to this function and expect update structure to be allocated and initialised to the given information, eg: modify_type is passed similarly as an argument to this function. Also this prevents from forgetting to set prepare_type when using __wt_update_alloc() and forces you to put a thought about what the prepare_state should be, even when default.

Reply:
prepare is the state of the transaction, which is inherited by update. May be later we can get away with prepare_state inside update structure and might only maintain at txn level. In my opinion, prepare state we can consider in conjunction with txnid. We will discuss.